platform/cxxsupport/mstd_memory

```diff
@@ -0,0 +1,616 @@
+
+// [unique.ptr]
+namespace impl
+{
+    /* Base version - use T * */
+    template <typename T, typename D, typename = void>
+    struct unique_ptr_type_helper {
+        typedef T *type;
+    };
+
+    /* if "remove_reference_t<D>::pointer" is a type, specialise to use it */
+    template <typename T, typename D>
+    struct unique_ptr_type_helper<T, D, mstd::void_t<typename remove_reference_t<D>::pointer>> {
+        typedef typename remove_reference_t<D>::pointer type;
+    };
+
+    template <class T, class D>
+    using unique_ptr_type_helper_t = typename unique_ptr_type_helper<T, D>::type;
+
+    // Want to eliminate storage for the deleter - could just use it as a base
+    // class, for empty base optimisation, if we knew it was a class. But it could be
+    // a pointer or reference. Here's a version that uses deleter as base,
+    template<class D, typename = void>
+    class deleter_store : private D {
+    public:
+        constexpr deleter_store() noexcept = default;
+        template <typename _D>
+        constexpr deleter_store(_D &&d) noexcept : D(std::forward<_D>(d)) { }
+
+        D &get_deleter() noexcept { return static_cast<D &>(*this); }
+        const D &get_deleter() const noexcept { return static_cast<const D &>(*this); }
+    };
+
+    //Here's a version that stores (for pointer/reference)
+    template<class D>
+    class deleter_store<D, enable_if_t<!is_class<D>::value>> {
+        D d;
+    public:
+        constexpr deleter_store() noexcept : d() { }
+        template <typename _D>
+        constexpr deleter_store(_D &&d) noexcept : d(std::forward<_D>(d)) { }
+
+        D &get_deleter() noexcept { return d; }
+        const D &get_deleter() const noexcept { return d; }
+    };
+}
+
+// [unique.ptr.dltr.dflt]
+template<class T>
+struct default_delete {
+    constexpr default_delete() noexcept = default;
+
+    template <class U, class = enable_if_t<is_convertible<U *, T *>::value>>
+    default_delete(const default_delete<U> &d) noexcept { }
+
+    void operator()(T *ptr) const
+    {
+        // Program is ill-formed if T is incomplete - generate diagnostic by breaking compilation
+        // (Behaviour of raw delete of incomplete class is undefined if complete class is non-trivial, else permitted)
+        static_assert(sizeof(T) == sizeof(T), "Cannot delete incomplete type");
+        delete ptr;
+    }
+};
+
+// [unique.ptr.dltr.dflt1]
+template<class T>
+struct default_delete<T[]> {
+    constexpr default_delete() noexcept = default;
+
+    template <class U, class = enable_if_t<is_convertible<U (*)[], T (*)[]>::value>>
+    default_delete(const default_delete<U> &d) noexcept { }
+
+    template <class U, class = enable_if_t<is_convertible<U (*)[], T (*)[]>::value>>
+    void operator()(U *ptr) const
+    {
+        delete[] ptr;
+    }
+};
+
+// [unique.ptr.single]
+template<
+    class T,
+    class D = default_delete<T>
+> class unique_ptr : public impl::deleter_store<D>
```

Comment:
Glad to see that class getting in.

Reply:
Yes, but sadly in use, it increases code size, at least in my tests.

You'd think it was "free", but the compiler doesn't seem to be as good at optimising out shuffling. For example, given:

     if (!ptr) {
         ptr = make_unique<Thing>(whatever);
     }

I've seen ARMC5 at least put in a pointless delete-of-current-pointer-contents on the assignment (but we just tested that the pointer is currently empty), and then a pointless delete of the temporary returned by make_unique (we know we just moved out of it).

Avoiding make_unique helped a bit, but still not as good as doing it by hand.

Maybe other compilers (and their std::unique_ptr) do better?

Reply:
Much as I want to believe the C++ guys' claims that C++ is as efficient as C, I keep seeing cases like this where it's just a bit too hard for the compiler to totally eliminate the generic stuff, and it does get bigger. At this embedded scale of tens of kilobytes of ROM and RAM, the impact is real...

Lots of little micro things you wouldn't think of.

Plus the annoyances like what I hit with `SingletonPtr` here - took me a number of hours to understand the initialization rules enough to comprehend the "mixed zero and constant initialization" problem I was hitting and deal with it to get them back out of the final link.

Was watching this while debugging that one: https://www.youtube.com/watch?v=D7Sd8A6_fYU

Made me a little grumpy - yes, there are misconceptions about C++ for embedded use, but it does also have real issues. And yes, sure, stronger type checking can move more debugging to compilation time, but often that debugging is much harder...

Reply:
It would be sweet if our documentations were explaining which C++ features is safe to use and what are the impacts on code size/speed (I'd love to write that). I hope most of the code won't turn into crazy template so debug ability should be okay; too bad that compilers doesn't have attributes to treat constructors as _literals_. It would be really useful for basic building blocks such as `Callback`. 

> Much as I want to believe the C++ guys' claims that C++ is as efficient as C, I keep seeing cases like this where it's just a bit too hard for the compiler to totally eliminate the generic stuff, and it does get bigger. At this embedded scale of tens of kilobytes of ROM and RAM, the impact is real...

I'd be interested to measure the impact on a real world application. Not that I doubt there is one (there will be one as long as there's no destructive moves) but an accurate picture would help us create guidelines to our community of users and internal developers.